src/wayback/_http.py

```diff
@@ -0,0 +1,591 @@
+
+
+class RetryAndRateLimitAdapter(WaybackHttpAdapter):
+    """
+    Adds rate limiting and retry functionality to an HTTP adapter. This class
+    can't actually make HTTP requests and should usually be used in a
+    multiple-inheritance situation. Alternatively, you can override the
+    ``request_raw()`` method.
+
+    Parameters
+    ----------
+    retries : int, default: 6
+        The maximum number of retries for failed HTTP requests.
+    backoff : int or float, default: 2
+        Number of seconds from which to calculate how long to back off and wait
+        when retrying requests. The first retry is always immediate, but
+        subsequent retries increase by powers of 2:
+
+            seconds = backoff * 2 ^ (retry number - 1)
+
+        So if this was `4`, retries would happen after the following delays:
+        0 seconds, 4 seconds, 8 seconds, 16 seconds, ...
+    timeout : int or float or tuple of (int or float, int or float), default: 60
+        A timeout to use for all requests.
+        See the Requests docs for more:
+        https://docs.python-requests.org/en/master/user/advanced/#timeouts
+    rate_limits : dict of (str, RateLimit)
+        The rate limits that should be applied to different URL paths. The keys
+        are URL path prefixes, e.g. ``"/cdx/search/"``, and the values are
+        :class:`wayback.RateLimit` objects. When requests are made, the rate
+        limit from the most specific matching path is used and execution will
+        pause to ensure the rate limit is not exceeded.
+
+
+    Examples
+    --------
+
+    Usage via multiple inheritance:
+
+    >>> class CombinedAdapter(RetryAndRateLimitAdapter, SomeActualHttpAdapterWithoutRateLimits):
+    >>>     ...
+
+    Usage via override:
+
+    >>> class MyHttpAdapter(RetryAndRateLimitAdapter):
+    >>>     def request_raw(
+    >>>         self,
+    >>>         method: str,
+    >>>         url: str,
+    >>>         *,
+    >>>         params: dict = None,
+    >>>         headers: dict = None,
+    >>>         follow_redirects: bool = True,
+    >>>         timeout: Union[int, Tuple[int, int]] = None
+    >>>     ) -> WaybackHttpResponse:
+    >>>         response = urllib.urlopen(...)
+    >>>         return make_response_from_urllib(response)
+    """
+
+    rate_limits: Dict[str, RateLimit]
+
+    # It seems Wayback sometimes produces 500 errors for transient issues, so
+    # they make sense to retry here. Usually not in other contexts, though.
+    retryable_statuses = frozenset((413, 421, 500, 502, 503, 504, 599))
+
+    # XXX: Some of these are requests-specific and should move WaybackRequestsAdapter.
+    retryable_errors = (ConnectTimeoutError, MaxRetryError, ReadTimeoutError,
+                        ProxyError, RetryError, Timeout)
+    # Handleable errors *may* be retryable, but need additional logic beyond
+    # just the error type. See `should_retry_error()`.
+    handleable_errors = (ConnectionError,) + retryable_errors
```

Comment:
These almost certainly need to move to either `RequestsAdapter` or `WaybackRequestsAdapter` in order for this to stay generic.